### PR TITLE
+per #18192 leveldb impl of EventsByTag query

### PR DESCRIPTION
**Pull request by patriknw**
- also refactoring of EventsByPersistenceIdPublisher

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #18192

**Comment by patriknw** [comment]
ping @ktoso

File: akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/EventsByPersistenceIdPublisher.scala [hunk 1]
```diff
@@ -88,45 +87,105 @@ class EventsByPersistenceIdPublisher(persistenceId: String, fromSequenceNr: Long
...
+      context.stop(self)
+  }
+
+  def receiveRecoverySuccess(highestSeqNr: Long): Unit
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class LiveEventsByPersistenceIdPublisher(
```
Comment (ktoso) [inline akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/EventsByPersistenceIdPublisher.scala]: ah I see, makes sense.

File: akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/LeveldbReadJournal.scala [hunk 2]
```diff
@@ -43,12 +47,20 @@ class LeveldbReadJournal(system: ExtendedActorSystem, config: Config) extends sc
   }
 
   def allPersistenceIds(hints: Seq[Hint]): Source[String, Unit] = {
+    // no polling for this query, the write journal will push all changes, but
+    // we still use the `NoRefresh` hint as user API
```
Comment (ktoso) [inline akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/LeveldbReadJournal.scala]: :+1: yep, good to keep the comment around

File: akka-persistence-query/src/test/scala/akka/persistence/query/journal/leveldb/EventsByTagSpec.scala [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  val colors = Set("green", "black", "blue")
+  override def toJournal(event: Any): Any = event match {
+    case s: String ⇒
+      var tags = colors.foldLeft(Set.empty[String])((acc, c) ⇒ if (s.contains(c)) acc + c else acc)
+      if (tags.isEmpty) event
+      else Tagged(event, tags)
+    case _ ⇒ event
+  }
+
+  override def fromJournal(event: Any, manifest: String): EventSeq = EventSeq.single(event)
```
Comment (ktoso) [inline akka-persistence-query/src/test/scala/akka/persistence/query/journal/leveldb/EventsByTagSpec.scala]: // PR unrelated: btw each time I see this it makes me want to go back and try to make `EventSeq` a value class somehow ;-)

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbJournal.scala [hunk 4]
```diff
@@ -21,12 +25,36 @@ private[persistence] class LeveldbJournal extends { val configPath = "akka.persi
...
+                  replyTo.tell(ReplayedTaggedMessage(adaptedPersistentRepr, tag, offset), Actor.noSender)
+                }
+            }.map(_ ⇒ highSeqNr)
+          }
+        }.map {
+          highSeqNr ⇒ RecoverySuccess(highSeqNr)
+        }.recover {
+          case e ⇒ ReplayMessagesFailure(e)
+        }.pipeTo(replyTo)
+
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbJournal.scala]: great that so much can be reused

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbJournal.scala [hunk 5]
```diff
@@ -43,18 +71,31 @@ private[persistence] object LeveldbJournal {
...
    */
-  case object SubscribeAllPersistenceIds extends SubscriptionCommand
-  case class CurrentPersistenceIds(allPersistenceIds: Set[String]) extends DeadLetterSuppression
-  case class PersistenceIdAdded(persistenceId: String) extends DeadLetterSuppression
+  final case object SubscribeAllPersistenceIds extends SubscriptionCommand
+  final case class CurrentPersistenceIds(allPersistenceIds: Set[String]) extends DeadLetterSuppression
+  final case class PersistenceIdAdded(persistenceId: String) extends DeadLetterSuppression
+
+  /**
+   * Subscribe the `sender` to changes (appended events) for a specific `tag`.
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbJournal.scala]: Mention the `Tagged` here? So people know how to "tag"?
Some docs will need to explain it anyway, tomorrow though I guess?

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/Tagged.scala [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * The LevelDB journal supports tagging of events that are used
+ * by the `EventsByTag` query. To specify the tags you create an
+ * [[akka.persistence.journal.EventAdapter]] that wraps the events
+ * in a `Tagged` with the given `tags`.
+ *
+ * The journal will unwrap the event and store the `payload`.
+ */
+case class Tagged(payload: Any, tags: Set[String])
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/Tagged.scala]: As discussed previously, agree on keeping it here for now, it's "for our journal".
We'll see if there's a need to provide a common one or not.

**Comment by ktoso** [comment]
LGTM

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbJournal.scala (hunk 5 above)
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbJournal.scala]: good point, docs will come tomorrow

**Comment by patriknw** [comment]
great, then I'll add docs in a separate PR

**Pull request merged**